File: 3/3.hs
primes = sieve [2..]
    where sieve (p:xs) =
              p : sieve [x | x <- xs, x `mod` p /= 0]

under n = takeWhile (\x -> x < n) primes

factors n = [x | x <- [2..n], n `mod` x == 0]

primeFactors n = [x | x <- (factors n), elem x (under (x+1))]

result = primeFactors 600851475143

main = print (maximum result)

File: 3/3.rs

fn main() {
    let num = 600851475143;
    let factors = prime_factors(num);
    let max = factors.iter().max();
    println!("{:?}", max);
}

fn sieve(bound: u64) -> Vec<u64> {
    // start from 2
    let mut is_prime = vec![true; bound as usize + 1];
    is_prime[0] = false; is_prime[1] = false;
    for i in 2..bound {
        if is_prime[i as usize] {
            let mut j = i + i;
            while j <= bound {
                is_prime[j as usize] = false;
                j += i;
            }
        }
    }
    is_prime.iter().enumerate().filter_map(|(prime, &is_prime)| if is_prime { Some(prime as u64) } else { None }).collect::<Vec<_>>()
}

fn prime_factors(n: u64) -> Vec<u64> {
    let primes = sieve((n as f64).sqrt() as u64 + 1 );

    primes.iter().filter(|&prime| n % prime == 0).map(|&x| x).collect::<Vec<_>>()
}
